### PR TITLE
Adding React state to markdown parser and save functinality

### PR DESCRIPTION
**Pull request by farhaanbukhsh**
This PR helps to convert the react state to markdown and send the data to edx-platform. This is required to show or have a markdown format of the  Problems.

~~**Discussions**: Link to any public dicussions about this PR or the design/architecture. Otherwise omit this.~~

~~**Dependencies**: None~~

~~**Screenshots**: Always include screenshots if there is any change to the UI.~~

~~**Sandbox URL**: TBD - sandbox is being provisioned.~~

~~**Merge deadline**: "None" if there's no rush, "ASAP" if it's critical, or provide a specific date if there is one.~~

**Testing instructions**:

1. Set up the devstack as mentioned [here](https://github.com/edx/frontend-lib-content-components/pull/110#issue-1360810689)
2. There are tests written for the react state parser which you can have a look and run using `npm test -- src/editors/containers/ProblemEditor/data/ReactStateParser.test.js --collectCoverage=false`
3. From studio if we are entering to the MFE and if any change is made and saved, this just reflects in the markdown and not in OLX so xblock preview doesn't have any change.
4. Only when you disable the flags and click edit you can see the change in the markdown editor.
5. You can also put a `console.log(content)` [here](https://github.com/open-craft/frontend-lib-content-components/pull/2/files#diff-c2d3d9d3c28ca1b4aa77706aa0aefb06b4d2f0facbb3a90e96a67963ed656eb3R122) for Gallery and similar place in api.js for live, you need to do a persist log here.

**Author notes and concerns**:

1. OLX has to be updated which will be done in the next tickets
2. We have still to find out how to parse and make the difference between question, description and label.

**Reviewers**
- [ ] @navinkarkera

Fixes: https://tasks.opencraft.com/browse/BB-6668
Co-authored-by: <>
Private-ref: <>

### CODE REVIEW

**Comment by navinkarkera** [comment]
Initial review:
Screenshot_20220915_133429

**Comment by farhaanbukhsh** [comment]
> Initial review: Screenshot_20220915_133429

I have already added https://github.com/open-craft/frontend-lib-content-components/pull/2/files#diff-6a1fb3ea3560ddd90d64c5a7a704221c15242033e7234accc61ba1202e8c11a4R178

a TODO to support selected and unselected feedback, this is the markdown parser from capa problem visual editor, although we should fix it here, but a thought of using OLX instead of Markdown to parse to the editor makes more sense now, so this markdown parser can entirely be depricated.

This might be out of the scope of the current ticket as well.

File: src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx [hunk 1]
```diff
@@ -1,14 +1,38 @@
...
+      dispatch(actions.problem.updateQuestion(content));
+    }
+  }
+  if (!refReady) { return null; }
   return (
     <div>
       <div>
         <h1>
-          QuestionWidget
+          Question
```
Comment (navinkarkera) [inline src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx]: Nit: missing internalization.

File: src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx [hunk 2]
```diff
@@ -1,14 +1,38 @@
...
+  const dispatch = useDispatch();
+  const { editorRef, refReady, setEditorRef } = hooks.prepareEditorRef();
+  const updateQuestionState = (dispatch, editorRef) => {
+    return () => {
+      // Only the question is recorded so everything becomes a question
+      // TODO: The support for lable and description and normal text.
+      let content = editorRef.current.getContent({format: 'text'});
+      dispatch(actions.problem.updateQuestion(content));
+    }
+  }
```
Comment (navinkarkera) [inline src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx]: Nit: can be replaced with `mapStateToProps` and `mapDispatchToProps`, https://github.com/open-craft/frontend-lib-content-components/blob/ce12d13552b5023afc7f0c8d7776bef1c65a2068/src/editors/containers/ProblemEditor/components/EditProblemView/AnswerWidget/AnswersContainer.jsx#L49-L55

File: src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx [hunk 3]
```diff
@@ -6,18 +6,22 @@ import QuestionWidget from './QuestionWidget';
...
-import { Editor } from '@tinymce/tinymce-react';
-import { ToggleButton, ToggleButtonGroup } from '@edx/paragon';
+import { ReactStateParser } from '../../data/ReactStateParser';
+
+
 
 export const EditProblemView = () => {
   const problemType = useSelector(selectors.problem.problemType);
-  const question = useSelector(selectors.problem.question);
+  const problemState = useSelector(selectors.problem.completeState);
```
Comment (navinkarkera) [inline src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx]: Nit: same here, can be replaced with `mapStateToProps`

**Review by navinkarkera — COMMENTED** [review]
Some small suggestions. Feel free to ignore them if you feel it is not important

File: src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx (hunk 3 above)
Comment (farhaanbukhsh) [reply src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx]: thanks :)

File: src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx (hunk 2 above)
Comment (farhaanbukhsh) [reply src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx]: Done

File: src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx (hunk 1 above)
Comment (farhaanbukhsh) [reply src/editors/containers/ProblemEditor/components/EditProblemView/QuestionWidget/index.jsx]: Done

File: src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx [hunk 4]
```diff
@@ -5,21 +5,31 @@ import SettingsWidget from './SettingsWidget';
...
   );
 };
 
-export default EditProblemView;
+export const mapStateToProps = (state) => ({
+  problemType: selectors.problem.problemType(state),
+  problemState: selectors.problem.completeState(state),
+})
+
+export default injectIntl(connect(mapStateToProps)(EditProblemView));
```
Comment (navinkarkera) [inline src/editors/containers/ProblemEditor/components/EditProblemView/index.jsx:35]: `injectIntl` probably not needed here. It is mostly used if you need to inject `intl` into props.

https://github.com/open-craft/frontend-lib-content-components/blob/476c450e6c6357e34ddf7c011c6e2237345db430/src/editors/containers/TextEditor/index.jsx#L133-L149

**Review by navinkarkera — APPROVED** [review]
:+1:

- [x] I tested this: Tested saving problems from mfe and reloading.
- [x] I read through the code
- [ ] I checked for accessibility issues
- [ ] Includes documentation
- [ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

**Pull request merged**